pkg/api/login_oauth.go

```diff
@@ -64,13 +70,18 @@ func (hs *HTTPServer) OAuthLogin(ctx *m.ReqContext) {
 		return
 	}
 
-	savedState, ok := ctx.Session.Get(session.SESS_KEY_OAUTH_STATE).(string)
-	if !ok {
+	savedState := ctx.GetCookie(OauthStateCookieName)
+
+	// delete cookie
+	ctx.Resp.Header().Del("Set-Cookie")
+	hs.writeOauthStateCookie(ctx, "", -1)
+
+	if savedState == "" {
 		ctx.Handle(500, "login.OAuthLogin(missing saved state)", nil)
 		return
 	}
 
-	queryState := ctx.Query("state")
+	queryState := hashStatecode(ctx.Query("state"), setting.OAuthService.OAuthInfos[name].ClientSecret)
```

Comment:
Im not sure if we need that extra seed or if state + secretkey is enough.

Reply:
Maybe use the equivalent of SetSuperSecureCookie from macaron - that's what we're using today for the remember me cookie. It's a shame though that macaron only offers the SetCookie with `others ...interface{}` arguments...

Reply:
Since this is only used for one way hashing I think sha256 is good enough. Let's revisit this if we add support for encrypting/decrypting content in cookies?

Reply:
Sure thing